src/serac/physics/heat_transfer.hpp

```diff
@@ -285,6 +285,64 @@ class HeatTransfer<order, dim, Parameters<parameter_space...>, std::integer_sequ
     cycle_ += 1;
   }
 
+  /**
+   * @brief Functor representing the integrand of a thermal material.  Material type must be
+   * a functor as well.
+   */
+  template <typename MaterialType>
+  struct ThermalMaterialIntegrand {
+    /**
+     * @brief Construct a ThermalMaterialIntegrand functor with material model of type `MaterialType`.
+     * @param[in] material A functor representing the material model.  Should be a functor, or a class/struct with
+     * public operator() method.  Must NOT be a generic lambda, or serac will not compile due to static asserts below.
+     */
+    ThermalMaterialIntegrand(MaterialType material) : material_(material) {}
+
+    // Due to nvcc's lack of support for generic lambdas (i.e. functions of the form
+    // auto lambda = [](auto) {}; ), this cannot be allowed.  In order for this code
+    // to be portable to CUDA platforms, the asserts below prevent serac from compiling
+    // if such a lambda is supplied.
+  private:
+    class DummyArgumentType {};
+    static_assert(!std::is_invocable<MaterialType, DummyArgumentType&>::value);
+    static_assert(!std::is_invocable<MaterialType, DummyArgumentType*>::value);
+    static_assert(!std::is_invocable<MaterialType, DummyArgumentType>::value);
```

Comment:
It's not clear to me what this is checking. If we want to detect for generic extended lambdas, then we have a compiler specific intrinsic:

https://gist.github.com/samuelpmish/5679a488c8724b215e96e107569b4fe9#file-main-cu-L49

Reply:
this static assert will pass if given a functor but fail if given a generic extended lambda, you can try it out in godbolt if you want.  This is not compiler specific, which in my opinion is better in this case because we want the option to be able to compile exclusively with clang or gcc

Reply:
> this static assert will pass if given a functor but fail if given a generic extended lambda, you can try it out in godbolt if you want. 

it incorrectly disallows functors with generic `operator()` definitions: https://godbolt.org/z/7nhEc1jYY

it incorrectly disallows non-extended lambdas: https://godbolt.org/z/rKrTcovcj

NVIDIA wrote a dedicated compiler intrinsic for checking exactly what we want to know, why not use it?

-----------

> This is not compiler specific, which in my opinion is better 

It's important to remember that the underlying limitation here is compiler specific, so imposing the restriction on all compilers isn't a benefit.

For instance, a user who wants to run their simulations on a CPU can (and probably should) use generic lambdas for their material / traction definitions, where appropriate. It doesn't seem right to take away a useful feature from _all_ users, because one compiler doesn't support it.

------------

> we want the option to be able to compile exclusively with clang or gcc

Then we can guard the assertion behind conditional compilation flags that are only enabled for the compiler that cares

```
    #if __NVCC__
    static_assert(!__nv_is_extended_device_lambda_closure_type(MaterialType), "error: extended lambdas not supported on nvcc");
    #endif
```

see: https://godbolt.org/z/9n9oxWbeb

Reply:
> NVIDIA wrote a dedicated compiler intrinsic for checking exactly what we want to know, why not use it?
If we guard the static assert behind an ifdef, users will be able to write generic lambdas for CPU versions of serac.  Then, if they want to switch to using GPU-enabled serac, their code will require refactoring.  In my opinion it is better to simplify the requirements to users, and force them to write templated functors.  I suppose the generic functor error is a limitation, but I feel that putting the static assert behind the if guard will only complicate things for users and make serac seem less portable. 

Ultimately, we want serac to be portable to CPU and GPU platforms.  We want user code to be portable to CPU and GPU platforms as well.  By forcing users to write templated functors, we are providing users with simplicity and guarantee of portability.

Reply:
I just spoke offline with @jamiebramwell and she said that still allowing generic lambdas for CPU generic lambdas would permit faster development/be more convenient for users.  I think this is a great point so we will go for @samuelpmishLLNL 's approach

Reply:
We should document this difference very thoroughly in Doxygen and in Sphinx. Maybe even providing an example on how to convert from generic lambdas to the functor.

We should also be very careful in our unit testing. Defaulting to the non-lambda version should be the normal, but also add a couple test cases to ensure the generic CPU lambdas continue to work.

Reply:
Just commenting here that I agree we should allow generic lambdas for rapid prototyping on non-nvcc compilers. We should just make sure that the vast majority of our examples and tests use the functor syntax as @white238 states.